src/parser/core/modules/BuffWindow.tsx

```diff
@@ -0,0 +1,329 @@
+import {MessageDescriptor} from '@lingui/core'
+import {t} from '@lingui/macro'
+import {Plural, Trans} from '@lingui/react'
+import {ActionLink} from 'components/ui/DbLink'
+import {RotationTable, RotationTableTargetData} from 'components/ui/RotationTable'
+import ACTIONS from 'data/ACTIONS'
+import {ActionData, getDataBy, StatusData} from 'data/getDataBy'
+import {BuffEvent, CastEvent} from 'fflogs'
+import _ from 'lodash'
+import Module, {dependency} from 'parser/core/Module'
+import GlobalCooldown from 'parser/core/modules/GlobalCooldown'
+import Suggestions, {TieredSuggestion} from 'parser/core/modules/Suggestions'
+import Timeline from 'parser/core/modules/Timeline'
+import React from 'react'
+
+class BuffWindowState {
+	start: number
+	end?: number
+	rotation: CastEvent[] = []
+
+	constructor(start: number) {
+		this.start = start
+	}
+
+	get gcds(): number {
+		return this.rotation
+			.map(e => getDataBy(ACTIONS, 'id', e.ability.guid) as TODO)
+			.filter(a => a && a.onGcd)
+			.length
+	}
+
+	getActionCountByIds(actionsById: number[]): number {
+		return this.rotation
+			.filter(e => actionsById.includes(e.ability.guid))
+			.length
+	}
+}
+
+interface SeverityTiers {
+	[key: number]: number
+}
+
+interface BuffWindowExpectedGCDs {
+	expectedPerWindow: number
+	suggestionContent: JSX.Element | string
+	severityTiers: SeverityTiers
+}
+
+interface BuffWindowRequiredGCDs {
+	iconAction: ActionData
+	actions: ActionData[]
+	suggestionContent: JSX.Element | string
+	severityTiers: SeverityTiers
+}
+
+interface BuffWindowTrackedCooldown {
+	action: ActionData
+	expectedPerWindow: number,
+	suggestionContent: JSX.Element | string
+	severityTiers: SeverityTiers
+}
+
+export default class BuffWindowModule extends Module {
```

Comment:
For new modules, esp. those in TS, I wanna start preferencing named exports (`export class` rather than `export default class`).

---

Given the below required `buffAction` and such, would it make sense for this to be an `abstract class`? That way you can enforce that subclasses provide it (unless they're JS) - the compiler will refuse to complete until the abstracts are implemented.

Reply:
See also the notes below - I had fiddled both ways with this being an abstract class... changed it to non-abstract to try to fix putting the filters on buffStatus in the init() and then never changed it back when filtering on buffStatus still didn't work in the parent init.  Will change back to abstract since I didn't gain anything by moving away from abstract.